src/NuGet.Clients/NuGet.SolutionRestoreManager/SolutionRestoreWorker.cs

```diff
@@ -251,6 +254,18 @@ public bool Restore(SolutionRestoreRequest request)
                         // Initialize if not already done.
                         await InitializeAsync();
 
+#if VS15
+                        if (request.RestoreSource == RestoreOperationSource.OnBuild)
+                        {
+                            if (!await SolutionManager.SolutionHasDeferredProjectsAsync()
+                                && SolutionManager.GetNuGetProjects().All(p => p is CpsPackageReferenceProject))
```

Comment:
If there are zero projects it looks like this new code path would be used. Which could bring some minor behavior changes. Maybe an explicit project count check should be done here, or return separately for that reason.